src/Illuminate/Database/Eloquent/Relations/HasOne.php

```diff
@@ -77,12 +76,22 @@ public function match(array $models, Collection $results, $relation)
      */
     protected function getDefaultFor(Model $model)
     {
+        $result = null;
+
         if (is_callable($this->withDefault)) {
-            return call_user_func($this->withDefault);
+            $result = call_user_func($this->withDefault);
```

Comment:
1. Why would the callback want to return `null`? That's a little weird.
2. Wouldn't it be better to pass a new linked-up model to the callback?

Maybe something like this:

```php
protected function getDefaultFor(Model $model)
{
    if (! $this->withDefault) {
        return;
    }

    $instance = $this->related->newInstance()->setAttribute(
        $this->getPlainForeignKey(), $model->getAttribute($this->localKey)
    );

    if (is_callable($this->withDefault)) {
        return call_user_func($this->withDefault, $instance);
    }

    if (is_array($this->withDefault)) {
        $instance->forceFill($this->withDefault);
    }

    return $instance;
}
```